### PR TITLE
Fix random number generation from cuRAND normal/log_normal distributions.

### PR DESCRIPTION
**Pull request by chantera**
This branch avoids exceptions caused by `CURAND_STATUS_LENGTH_NOT_MULTIPLE` error that occurs when an odd number of elements is specified for random number generation from cuRAND normal/log_normal distributions.
See the following document: https://docs.nvidia.com/cuda/curand/group__HOST.html#group__HOST_1gb9280e447ef04e1dec4611720bd0eb69

### CODE REVIEW

File: primitiv/core/device.h [hunk 1]
```diff
@@ -335,6 +335,14 @@ class Device
     return x.mutable_handle();
   }
 
+  /**
+   * Returns the allocated memory size of a Tensor.
+   * @return Memory size.
+   */
+  static std::size_t allocated_size(const Tensor &x) {
```
Comment (odashi) [inline primitiv/core/device.h]: It looks not necessary if `Tensor` provides the same public function.

File: primitiv/core/memory_pool.cc [hunk 2]
```diff
@@ -60,6 +61,14 @@ std::shared_ptr<void> MemoryPool::allocate(std::size_t size) {
     supplied_.emplace(ptr, shift);
   }
 
+  if (mem_size < size) {
```
Comment (odashi) [inline primitiv/core/memory_pool.cc]: Is this condition necessary? `mem_size` is always largher than or equal to `size` according to `calculate_shifts`.

File: primitiv/core/tensor.h [hunk 3]
```diff
@@ -208,9 +213,16 @@ class Tensor {
    */
   void *mutable_handle();
 
+  /**
+   * Returns the size of the allocated memory to store values.
+   * @return Memory size.
+   */
+  std::size_t allocated_size() const { return allocated_size_; }
```
Comment (odashi) [inline primitiv/core/tensor.h]: Let's make this function public.

Comment (odashi) [inline primitiv/core/tensor.h]: Add test cases for this function.

File: primitiv/core/memory_pool.cc [hunk 4]
```diff
@@ -30,7 +30,7 @@ MemoryPool::~MemoryPool() {
   release_reserved_blocks();
 }
 
-std::shared_ptr<void> MemoryPool::allocate(std::size_t size) {
+std::shared_ptr<void> MemoryPool::allocate(std::size_t size, std::size_t * const allocated_size) {
```
Comment (odashi) [inline primitiv/core/memory_pool.cc:35]: Add test cases to check `allocated_size`.

**Review by odashi — CHANGES_REQUESTED** [review]

**Comment by chantera** [comment]
Thank you for your review.
I fixed all.

File: primitiv/devices/cuda/ops/random_log_normal.cu [hunk 5]
```diff
@@ -10,7 +10,7 @@ namespace devices {
 void CUDA::random_log_normal_impl(float mean, float sd, Tensor &y) {
   std::size_t size = y.shape().size();
   if (size % 2 != 0) {
-    std::size_t capacity = allocated_size(y) / sizeof(float);
+    std::size_t capacity = y.allocated_size() / sizeof(float);
     if (capacity <= size) {
```
Comment (odashi) [inline primitiv/devices/cuda/ops/random_log_normal.cu:14]: This check may always fail if the size is 1. Add some option to `MemoryPool` for allocating memories with a guarantee of minimum sizes (e.g., add `minimum_size` option to the constructor) is recommended.

File: test/cuda_memory_pool_test.cc [hunk 6]
```diff
@@ -103,6 +103,27 @@ TEST_F(MemoryPoolTest_CUDA, CheckAllocate) {
...
+  while (size < prop.totalGlobalMem) {
+     ASSERT_NO_THROW(pool.allocate(size, &allocated_size));
+     EXPECT_EQ(size, allocated_size);
+     size <<= 1;
+  }
+
+  EXPECT_GT(allocated_size, 0);
+  EXPECT_THROW(pool.allocate(size << 1, &allocated_size), Error);
+  EXPECT_EQ(0, allocated_size);
+}
```
Comment (odashi) [inline test/cuda_memory_pool_test.cc:167]: Add more tests which checks the cases when `allocated_size` is not equal to `size`. E.g., if we try to allocate memory with 3 bytes, `MemoryPool` may set 4 into `allocated_size`.

File: test/tensor_test.cc [hunk 7]
```diff
@@ -94,6 +94,35 @@ TEST_F(TensorTest, CheckNewMatrixMinibatchWithData) {
...
 }
 
+TEST_F(TensorTest, CheckAllocatedSize) {
+  vector<float> data;
+  for (std::uint32_t i = 1; i <= 10; ++i) {
+    data.emplace_back(2.0 * i);
+    for (Device *dev : devices) {
+      // NOTE(chantera):
+      // sizeof(half) is assumed to be sizeof(float) / 2.
+      const std::size_t required_memsize
```
Comment (odashi) [inline test/tensor_test.cc]: Use an additional variable for readability to store `sizeof(float or half)`.

File: test/tensor_test.cc [hunk 8]
```diff
@@ -94,6 +94,35 @@ TEST_F(TensorTest, CheckNewMatrixMinibatchWithData) {
   }
 }
 
+TEST_F(TensorTest, CheckAllocatedSize) {
```
Comment (odashi) [inline test/tensor_test.cc:114]: These checks should also be implemented in existing tests in this file.

**Review by odashi — DISMISSED** [review]
LGTM

**Review by odashi — APPROVED** [review]

**Pull request merged**